scipy/linalg/_linsolve.pyx.in

```diff
@@ -0,0 +1,603 @@
+    to asarray_chkfinite the performance gets better as the
+    array size increases. Starting from n=3 it is ~x2 faster
+    for finite arrays and faster if array has inf/nan due to
+    early exit capability.
+    """
+    cdef int n = <int>A.shape[0]
+    cdef int r, c
+    cdef lapack_t entry
+    if lapack_t in lapack_cz_t:
+        for r in xrange(n):
+            for c in xrange(n):
+                entry = A[r, c]
+                if not(isfinite(entry.real) and isfinite(entry.imag)):
+                    return False
+    else:
+        for r in xrange(n):
+            for c in xrange(n):
+                if (not isfinite(A[r, c])):
+                    return False
+    return True
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef inline bint _ones_on_diagonal(lapack_t[:, ::1] a) nogil:
+    cdef size_t n = len(a)
+    cdef int k
+    cdef lapack_t one = 1.0
+
+    for k in range(n):
+        if a[k, k] != one:
+            return False
+
+    return True
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef (int, int) _is_tri_c(lapack_t[:, ::1]A) nogil:
+    """
+    Computes the lower and the upper band width of a square
+    matrix. If both bands are zero, it's diagonal, if either
+    is zero then lower/upper triangular etc. If all entries
+    are full then for nxn matrix the result is (n-1, n-1).
+
+    This is the row-major layout version (C-contiguous).
+    """
+    cdef size_t n = A.shape[0]
+    cdef int r, c, lower_band, upper_band
+    lower_band = 0
+    upper_band = 0
+    for r in xrange(n):
+        for c in xrange(n):
+            # Only bother if outside the existing band:
+            if (A[r, c] != 0.):
+                if r > c and (r - c) > lower_band:
+                    lower_band = r - c
+                elif r < c and (c - r) > upper_band:
+                    upper_band = c - r
+
+    return lower_band, upper_band
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef (bint, bint) _is_sym_her_c(lapack_t[:, ::1]A) nogil:
+    """
+    Checks for the symmetry or Hermitianness for a square matrix.
+    For reals it doesn't matter however for complex matrices,
+    symmetry can still be exploited even though the matrix is not
+    Hermitian.
+    """
+    cdef size_t n = A.shape[0]
+    cdef int r, c
+    cdef lapack_t x, y
+    cdef bint symmetric = 1
+    cdef bint hermitian = 1
+
+    # Real branch
+    if lapack_t in lapack_sd_t:
+        for r in xrange(n):
+            for c in xrange(n):
+                if r == c:
+                    continue
+                else:
+                    if A[r, c] != A[c, r]:
+                        return 0, 0
+        return 1, 1
+    # Complex branch
+    else:
+        for r in xrange(n):
+            for c in xrange(n):
+                # look-up once
+                x, y = A[r, c], A[c, r]
+                # once caught no need to keep checking continue for sym
+                if hermitian:
```

Comment:
You'd have to do some benchmarking to see if this really helps, but you can avoid some if predictive branching and hopefully take some advantage of logic short circuits by doing the following
```python
for r in range(n):
    for c in range(n):
        x, y = A[r, c], A[c, r]
        symmetric |= (x != y)
        hermitian |= symmetric or (x != y.conjugate())
        if not (symmetric or hermition):
            return False, False
```

Reply:
Unless `symmetric != hermitian`, I guess I assumed for real-valued matrices they would be symmetric and hermitian, but hermitian is usually only a property of complex-valued matrices